src/llmq/quorums_chainlocks.cpp

```diff
@@ -119,88 +178,143 @@ void CChainLocksHandler::ProcessNewChainLock(const NodeId from, const llmq::CCha
             return;
         }
 
-        if (!bestChainLock.IsNull() && clsig.nHeight <= bestChainLock.nHeight) {
+        if (bestChainLockBlockIndex != nullptr && clsig.nHeight <= bestChainLockBlockIndex->nHeight) {
             // no need to process/relay older CLSIGs
             return;
         }
     }
 
-    const uint256 requestId = ::SerializeHash(std::make_pair(CLSIG_REQUESTID_PREFIX, clsig.nHeight));
-    if (!quorumSigningManager->VerifyRecoveredSig(Params().GetConsensus().llmqTypeChainLocks, clsig.nHeight, requestId, clsig.blockHash, clsig.sig)) {
-        LogPrint(BCLog::CHAINLOCKS, "CChainLocksHandler::%s -- invalid CLSIG (%s), peer=%d\n", __func__, clsig.ToString(), from);
-        if (from != -1) {
-            LOCK(cs_main);
-            Misbehaving(from, 10);
+    CBlockIndex* pindexSig;
+    CBlockIndex* pindexScan;
+    {
+        LOCK(cs_main);
+        pindexSig = pindexScan = LookupBlockIndex(clsig.blockHash);
+        if (pindexScan == nullptr) {
+            // we don't know the block/header for this CLSIG yet, try scanning quorums at chain tip
+            pindexScan = chainActive.Tip();
         }
-        return;
     }
 
-    CBlockIndex* pindex;
-    {
-        LOCK(cs_main);
-        pindex = LookupBlockIndex(clsig.blockHash);
+    const auto llmqTypeChainLocks = Params().GetConsensus().llmqTypeChainLocks;
+    const auto signingActiveQuorumCount = GetLLMQParams(llmqTypeChainLocks).signingActiveQuorumCount;
+    const auto quorums_scanned = llmq::quorumManager->ScanQuorums(llmqTypeChainLocks, pindexScan, signingActiveQuorumCount);
+
+    std::set<CQuorumCPtr> quorums_tried;
+    int nStep{0};
+    // First try default quorum
+    uint256 requestId = ::SerializeHash(std::make_pair(CLSIG_REQUESTID_PREFIX, clsig.nHeight));
+    while (true) {
+        LogPrint(BCLog::CHAINLOCKS, "CChainLocksHandler::%s -- CLSIG (%s) requestId=%s, peer=%d\n",
+            __func__, clsig.ToString(), requestId.ToString(), from);
+        auto quorum = CSigningManager::SelectQuorumForSigning(llmqTypeChainLocks, requestId, clsig.nHeight);
+        if (quorum == nullptr) {
+            return;
+        }
+        if (quorums_tried.insert(quorum).second) {
+            uint256 signHash = CLLMQUtils::BuildSignHash(llmqTypeChainLocks, quorum->qc.quorumHash, requestId, clsig.blockHash);
+            if (clsig.sig.VerifyInsecure(quorum->qc.quorumPublicKey, signHash)) {
+                LOCK(cs);
+                // found valid
+                auto it = bestChainLockCandidates.find(clsig.nHeight);
```

Comment:
you can use try_emplace if you use c++14 or higher which will remove the extra lookup it will insert in place by forwarding arguments and give you back iterator to modify incase it exists and you want to update it.

Reply:
We still use c++14 but thanks, that's smth to refactor in the future I guess :)